squire_lib/src/ffi.rs

```diff
@@ -13,361 +18,346 @@ use crate::{
     },
     standard_scoring::{StandardScore, StandardScoring},
     swiss_pairings::SwissPairings,
-    tournament::{Tournament, TournamentPreset},
 };
+use dashmap::DashMap;
+use once_cell::sync::OnceCell;
+use serde_json;
+use std::ffi::CStr;
+use std::ffi::CString;
+use std::fs::{read_to_string, remove_file, rename, write};
+use std::option::Option;
+use std::os::raw::c_char;
+use std::ptr::null;
+use std::time::Duration;
+use std::vec::Vec;
+use uuid::Uuid;
+use std::alloc::{Allocator, System, Layout};
+use std::ptr;
+
+lazy_static! {
+    /// NULL UUIDs are returned on errors
+static ref NULL_UUID_BYTES:
+    [u8; 16] = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0];
+}
 
-use mtgjson::model::deck::Deck;
+/// A map of tournament ids to tournaments
+/// this is used for allocating ffi tournaments
+/// all ffi tournaments are always deeply copied
+/// at the lanuage barrier
+static FFI_TOURNAMENT_REGISTRY: OnceCell<DashMap<TournamentId, Tournament>> = OnceCell::new();
+const BACKUP_EXT: &str = ".bak";
+
+#[no_mangle]
+pub extern "C" fn init_squire_ffi() {
+    let map: DashMap<TournamentId, Tournament> = DashMap::new();
+    FFI_TOURNAMENT_REGISTRY.set(map);
+}
 
-use libc::c_char;
-use std::{
-    collections::HashMap,
-    ffi::{CStr, CString},
-    hash::{Hash, Hasher},
-    time::Duration,
-};
+/// Helper function for cloning strings
+unsafe fn clone_string_to_c_string(s: String) -> *mut c_char {
+    let len: usize = s.len() + 1;
+    let s_str = s.as_bytes();
+    
+    let ptr = System.allocate(Layout::from_size_align(len, 1).unwrap()).unwrap().as_mut_ptr() as *mut c_char;
+    let mut slice = &mut *(ptr::slice_from_raw_parts(ptr, len) as *mut [c_char]);
+    let mut i: usize = 0;
+    while i < s.len() {
+        slice[i] = s_str[i] as i8;
+        i += 1;
+    }
+    slice[i] = 0;
 
-impl Tournament {
-    /// Returns 0 if everything is ok.
-    /// Returns 1 if there is an error with the name conversion
-    /// Returns 2 if there is an error with the format conversion
-    #[allow(unused_assignments)]
+    return ptr;
+}
+
+/// TournamentIds can be used to get data safely from
+/// the Rust lib with these methods
+impl TournamentId {
+    /// Returns the name of a tournament
+    /// Returns NULL if an error happens
+    /// This is heap allocated, please free it
     #[no_mangle]
-    pub extern "C" fn from_preset_c(
-        mut expected: *mut Self,
-        name_buf: *mut c_char,
-        preset: TournamentPreset,
-        format_buf: *mut c_char,
-    ) -> usize {
-        let name_str = unsafe { CStr::from_ptr(name_buf) };
-        let name = match name_str.to_str() {
-            Ok(s) => s.to_string(),
-            Err(_) => {
-                return 1;
+    pub extern "C" fn tid_name(self: Self) -> *const c_char {
+        let tourn: Tournament;
+        match FFI_TOURNAMENT_REGISTRY.get().unwrap().get(&self) {
```

Comment:
This is an extremely small detail that is unlikely to be an issue on individual desktop clients. However, it could cause problems, so it's worth noting. Both `DashMap::get` and `DashMap::get_mut` can cause deadlocks. We might consider adding a super simple spin lock at some point.

Reply:
The ffi side isn't going to be thread safe so this isn't needed. 

Also a normal mutex is better than a spin lock Actually that's a hot take see what torvalds has to say on it.